### PR TITLE
System attributes md5 for sqs

### PR DESCRIPTION
**Pull request by mullermp**
Adds system attributes for verification

### CODE REVIEW

File: gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/md5s.rb [hunk 1]
```diff
@@ -83,10 +88,27 @@ def validate_attributes(attributes, response)
             end
           end
 
+          def validate_system_attributes(system_attributes, response)
+            calculated_md5 = md5_of_message_system_attributes(system_attributes)
+            returned_md5 = response.md5_of_message_system_attributes
+            if returned_md5 != calculated_md5
+              error_message =  mismatch_error_message(
```
Comment (jterapin) [inline gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/md5s.rb]: Extra space between `=` and `mistmatch_error_message`

File: gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/md5s.rb [hunk 2]
```diff
@@ -83,10 +88,27 @@ def validate_attributes(attributes, response)
...
+                'message system attributes',
+                calculated_md5,
+                returned_md5,
+                response)
```
Comment (jterapin) [inline gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/md5s.rb]: Closing brace need to be on the next line

File: gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/md5s.rb [hunk 3]
```diff
@@ -110,6 +132,30 @@ def md5_of_message_attributes(message_attributes)
...
+            encoded = { }
+            message_system_attributes.each do |name, attribute|
+              name = name.to_s
+              encoded[name] = String.new
+              data_type_without_label = DATA_TYPE.match(attribute[:data_type])[1]
+              encoded[name] << encode_length_and_bytes(name) <<
+              encode_length_and_bytes(attribute[:data_type]) <<
+              [TRANSPORT_TYPE_ENCODINGS[data_type_without_label]].pack('C'.freeze)
+
+              if attribute[:string_value] != nil
```
Comment (jterapin) [inline gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/md5s.rb]: Are we able to use `!attribute[:string_value].nil?` Same goes for line 147

File: gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/md5s.rb [hunk 4]
```diff
@@ -83,10 +88,27 @@ def validate_attributes(attributes, response)
...
+          end
+
           def md5_of_message_body(message_body)
             OpenSSL::Digest::MD5.hexdigest(message_body)
           end
 
+          # MD5 of Message Attributes and System Attributes are effectively
+          # the same calculation. However, keeping these as two methods because
+          # they are modeled as two different shapes.
+          ###
```
Comment (jterapin) [inline gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/md5s.rb:113]: What does `###` mean in comments?

File: gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/md5s.rb [hunk 5]
```diff
@@ -110,6 +132,30 @@ def md5_of_message_attributes(message_attributes)
...
 
+          def md5_of_message_system_attributes(message_system_attributes)
+            encoded = { }
+            message_system_attributes.each do |name, attribute|
+              name = name.to_s
+              encoded[name] = String.new
+              data_type_without_label = DATA_TYPE.match(attribute[:data_type])[1]
+              encoded[name] << encode_length_and_bytes(name) <<
+              encode_length_and_bytes(attribute[:data_type]) <<
+              [TRANSPORT_TYPE_ENCODINGS[data_type_without_label]].pack('C'.freeze)
```
Comment (alextwoods) [inline gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/md5s.rb]: probaby don't need to do `freeze` on the constant here

File: gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/md5s.rb (hunk 4 above)
Comment (mullermp) [reply gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/md5s.rb:113]: Logical separator

**Review by alextwoods — APPROVED** [review]
LGTM.

Nit: Update PR description

**Review by jterapin — APPROVED** [review]
LGTM!

**Pull request merged**